### PR TITLE
[FIX] web_editor: power box disappears in collaborative fix

### PR DESCRIPTION
**Pull request by deso-odoo**
**Current behavior before PR:**

- Power box uses Patience diff algorithm to check the text added by the user for
search, it previously used the entire editable for differences, which caused
the issue in collaborative. When one user type something on a different node,
the change would affect the editable and be considered in diff, causing updates
in other users powerbox.
- The Patience diff algorithm now uses the text added in the `closestBlock` node to
the node where the powerbox is opened. If the user switched to a different block
using `ArrowLeft` or `ArrowRight` keys, it would result in the algorithm not
searching for the typed text in the other block.

**Desired behavior after PR is merged:**

- Instead of checking the entire editable for the diff, we now check the current
block node where the power box was opened. This prevents scenarios where other
users powerbox would update when one user would type on different block node.
- The powerbox will be closed if the keyup event occurs in a different block than
the one in which the powerbox was initially opened.

task-3212128

### CODE REVIEW

File: addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js [hunk 1]
```diff
@@ -174,8 +175,11 @@ export class Powerbox {
         this.options.onActivate && this.options.onActivate();
         this._currentOpenOptions = openOptions;
 
+        const selection = this.options.document.getSelection();
         const openOnKeyupTarget =
             this._currentOpenOptions.openOnKeyupTarget || this.options.editable;
+        const currentNode =
+            this._currentOpenOptions.currentNode || this.options.editable;
```
Comment (shsa-odoo) [inline addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js:182]: @deso-odoo declare them where its needed, in your case it should be above `showOnceOnKeyup` as you started using selection, cuurentNode after this function.Its just a good practice to keep in mind. Cheers!

File: addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js [hunk 2]
```diff
@@ -194,13 +196,18 @@ export class Powerbox {
...
         const keyup = async ev => {
             if (ev.key === 'ArrowDown' || ev.key === 'ArrowUp') {
                 ev.preventDefault();
                 return;
             }
             if (!initialTarget) return;
+            // Update initialTarget in case of backspace.
+            if (!currentElement.contains(initialTarget)) {
+                initialTarget = selection.anchorNode;
+            }
```
Comment (Zynton) [inline addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js:210]: Same questions as in the 16 branch: why is this needed, and why specifically for backspace?

File: addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js [hunk 3]
```diff
@@ -1,6 +1,6 @@
 /** @odoo-module **/
 import { patienceDiff } from './patienceDiff.js';
-import { getRangePosition } from '../utils/utils.js';
+import { closestElement, getRangePosition } from '../utils/utils.js';
```
Comment (Zynton) [inline addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js:3]: Same remark as in the 16.0 branch regarding the commit message.

File: addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js [hunk 4]
```diff
@@ -301,15 +308,6 @@ export class Powerbox {
         if (this._currentOpenOptions.forceShow) showOnceOnKeyup();
     }
```
Comment (Zynton) [inline addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js:310]: I'm guessing this change is unrelated and the function was just unused. I wouldn't remove it though because it's a public API function and there might be customers who use it in their custom code.

**Review by Zynton — CHANGES_REQUESTED** [review]
@Goaman Could you take a look at this please? Your comment on the task implied that the issue was introduced by a commit I made in 16.0 and there's a PR in 16.0 to fix it just the way you said, but then there is also this PR in 15.0, which doesn't do what its commit message says and seems to bypass openOnKeyupTarget (while keeping its event listener). Does this make sense to you?

File: addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js (hunk 2 above)
Comment (deso-odoo) [reply addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js:210]: For backspace because on backspace the `initialTarget` which is `selection.anchorNode` is removed and it is no longer in `currentElement` so I reset it. In general case where `selection.anchorNode` is no longer present in `currentElement`.

**Comment by deso-odoo** [comment]
> @Goaman Could you take a look at this please? Your comment on the task implied that the issue was introduced by a commit I made in 16.0 and there's a PR in 16.0 to fix it just the way you said, but then there is also this PR in 15.0, which doesn't do what its commit message says and seems to bypass openOnKeyupTarget (while keeping its event listener). Does this make sense to you?

Here `openOnKeyupTarget` is ev.target which would be the `editable`, there still is `showOnceOnKeyup()` which is bind on `openOnKeyupTarget` (line [192](https://github.com/odoo/odoo/pull/126179/files#diff-d51c66003abbf352965691b6d10f5e2d8029c568fa0383558d480ee5564dc9c6R192)) which opens the powerbox.

File: addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js [hunk 5]
```diff
@@ -181,11 +182,12 @@ export class Powerbox {
...
                 : this._filter(term, this._currentOpenOptions.commands);
 
+        const selection = this.options.document.getSelection();
         const showOnceOnKeyup = () => {
             this.show();
             openOnKeyupTarget.removeEventListener('keyup', showOnceOnKeyup, true);
-            initialTarget = openOnKeyupTarget;
-            this._initialValue = openOnKeyupTarget.textContent;
+            initialTarget = selection.anchorNode;
+            this._initialValue = selection.anchorNode.textContent;
```
Comment (Zynton) [inline addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js:190]: Thanks for your explanation, it's now all clear to me. Now, what if `selection.anchorNode` is a text node that was split in pieces, like so: `"abc""[]d""ef"`? If we just type it seems ok but then say I press the right arrow key twicebefore typing some more (`"abc""/d""eCOMMAND[]f"`). What I typed here won't be taken into account in the diff right? Using `closestElement(selection.anchorNode)` might suffice but then we don't support the same case but where `e` is in a `<i>` tag for instance.
So I suggest using `closestBlock(selection.anchorNode)` instead.

cc @sebgeelen @Goaman for their opinions.

Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js:190]: `closestBlock` seems the right approach to me indeed. 
It would still need to be thoroughly tested to ensure it work as intended though.

**Comment by Zynton** [comment]
@deso-odoo If you could make the changes requested in the review, and add a few tests, we'll be good to go. Thanks for your work and clarifications.

File: addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js (hunk 5 above)
Comment (deso-odoo) [reply addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js:190]: Taking initialTarget as `closestBlock` how about we close the powerbox if user moves out of that block using arrow keys?

Comment (Zynton) [reply addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js:190]: That would make sense yes!

**Comment by deso-odoo** [comment]
> @deso-odoo If you could make the changes requested in the review, and add a few tests, we'll be good to go. Same for the 16.0 version. Thanks for your work and clarifications.

@Zynton Implemented requested changes!

File: addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js (hunk 3 above)
Comment (Zynton) [reply addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js]: This comment still applies. Please be more specific in your commit message so it's clear what functional issue your commit addresses.

File: addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js [hunk 6]
```diff
@@ -184,8 +184,10 @@ export class Powerbox {
         const showOnceOnKeyup = () => {
             this.show();
             openOnKeyupTarget.removeEventListener('keyup', showOnceOnKeyup, true);
-            initialTarget = openOnKeyupTarget;
-            this._initialValue = openOnKeyupTarget.textContent;
+            const selection = this.options.document.getSelection();
+            const currentBlock = closestBlock(selection.anchorNode) || this.options.editable;
```
Comment (Zynton) [inline addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js:188]: This should probably check that `selection` exists before getting `anchorNode` from it. From what I can see, `closestBlock` should return `undefined` if `selection.anchorNode` is not defined so that's ok.

```suggestion
            const currentBlock = (selection && closestBlock(selection.anchorNode)) || this.options.editable;
```

File: addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js [hunk 7]
```diff
@@ -207,7 +209,11 @@ export class Powerbox {
                 true,
             );
             this._lastText = diff.bMove.join('');
-            if (this._lastText.match(/\s/) && this._currentOpenOptions.closeOnSpace !== false) {
+            const selection = this.options.document.getSelection();
+            if (
+                (this._lastText.match(/\s/) && this._currentOpenOptions.closeOnSpace !== false) ||
+                initialTarget !== closestBlock(selection.anchorNode)
```
Comment (Zynton) [inline addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js:215]: Same here:
```suggestion
                !selection ||
                initialTarget !== closestBlock(selection.anchorNode)
```
This is a bit out of scope though, it's not part of the fix. Please make a new commit for it, along with its test.

**Review by Zynton — CHANGES_REQUESTED** [review]
Nice job, we're almost there!

File: addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js (hunk 3 above)
Comment (deso-odoo) [reply addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js]: Updated!

File: addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js [hunk 8]
```diff
@@ -4188,4 +4192,45 @@ X[]
             });
         });
     });
+    describe('powerbox', () => {
+        it('should not filter the powerbox contents when collaborator type on different block node', async () => {
+            await testEditor(BasicEditor, {
+                contentBefore: '<p>ab</p><p>c[]d</p>',
+                stepFunction: async (editor) => {
+                    await insertText(editor, '/');
+                    await insertText(editor, 'heading');
```
[the remark this answers is not in the document]
Comment (Zynton) [reply addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js:4201]: OK fine by me.

**Comment by Zynton** [comment]
Excellent work, thank you!
@robodoo rebase-ff r+
@fw-bot up to saas-15.2
